### PR TITLE
Kotlin Thrift Example

### PR DESCRIPTION
**Pull request by adomokos**
This PR is an example of how we could use `thrift` from Kotlin.
It's currently a simple, single-threaded server/client example, after talking with @wilsoniya, a multi-threaded example would be beneficial.

I'll remove the construction sign when it's ready for full review.

### CODE REVIEW

File: kotlin/thrift-example/client/src/main/kotlin/thriftclient/Client.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import fortune.VehicleDescription
+import org.apache.thrift.TException
+import org.apache.thrift.protocol.TBinaryProtocol
+import org.apache.thrift.transport.TSocket
+
+object Client {
+    @JvmStatic
+    fun main(_args: Array<String>) {
+        try {
+            val transport = TSocket("localhost", 9090)
```
Comment (JustinB-SH) [inline kotlin/thrift-example/client/src/main/kotlin/thriftclient/Client.kt]: I'll need to add this to the grpc example too.  You can add the `.use` receiver method to auto-close the TSocket.  https://www.baeldung.com/kotlin-try-with-resources

File: kotlin/thrift-example/model/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+plugins {
+    java
+}
+
+dependencies {
+    compile(files("../lib/javax.annotation-3.1.2.2.jar"))
```
Comment (JustinB-SH) [inline kotlin/thrift-example/model/build.gradle.kts]: Is the lib jar necessary, or is it possible to add it as a root-level dependency as I did in grpc-java? https://github.com/spothero/idl-test/blob/6c0a63cf85b5f2ca7b11e6b091cc27b55ba21da3/kotlin/grpc-java/model/build.gradle.kts#L26

Comment (adomokos) [reply kotlin/thrift-example/model/build.gradle.kts]: Thank you for suggesting this! I've tried to come up with this solution, couldn't find it. Works like a charm. 👍

File: kotlin/thrift-example/client/src/main/kotlin/thriftclient/Client.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        println("From client...")
+        val response = client.GetFortune(RequestFactory.create())
+        println("The response is:")
+        println(response)
+    }
+
+    object RequestFactory {
+        private fun createCar(): Car {
+            val car = Car()
+            car.make = "Trabant"
```
[the remark this answers is not in the document]
Comment (adomokos) [reply kotlin/thrift-example/client/src/main/kotlin/thriftclient/Client.kt]: `Trabant` is dear to my heart. I learned to drive on that beast.

File: kotlin/thrift-example/client/src/main/kotlin/thriftclient/Client.kt (hunk 1 above)
Comment (adomokos) [reply kotlin/thrift-example/client/src/main/kotlin/thriftclient/Client.kt]: Forgot about `use`, thanks for reminding me.

File: kotlin/thrift-example/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    apply(plugin = "java")
+    apply(plugin = "org.jlleitschuh.gradle.ktlint")
+    dependencies {
+        compile(kotlin("stdlib"))
+        compile("org.apache.thrift:libthrift:0.12.0")
+
+        // Used for logging
+        compile("org.jetbrains.kotlin:kotlin-reflect")
+        compile("ch.qos.logback:logback-core:1.2.3")
+        compile("ch.qos.logback:logback-classic:1.2.3")
```
Comment (adomokos) [inline kotlin/thrift-example/build.gradle.kts]: @JustinB-SH  - not sure if you knew this, but you can put project dependencies in the root `build.gradle.kts` file, and those can be leveraged in all subprojects. You might be able to do the same with all the proto dependencies in yours.

File: kotlin/thrift-example/server/src/main/kotlin/thriftserver/FortuneTellerService.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import fortune.FortuneTeller
+import kotlin.random.Random
+
+class FortuneTellerService : FortuneTeller.Iface {
+    override fun GetFortune(request: FortuneRequest): FortuneResponse {
+        return ResponseBuilder.buildResponse(request)
+    }
+
+    object ResponseBuilder {
+        fun buildResponse(request: FortuneRequest): FortuneResponse {
```
Comment (wilsoniya) [inline kotlin/thrift-example/server/src/main/kotlin/thriftserver/FortuneTellerService.kt]: Either here or as a followup you should make it possible for an `UnfortunateException` to be raised per the service definition. In my GRPC implementation I'm raising an exception when one of the finger_lengths is >= 100.

Comment (adomokos) [reply kotlin/thrift-example/server/src/main/kotlin/thriftserver/FortuneTellerService.kt]: That's a good idea, I'll add it.

**Comment by adomokos** [comment]
@JustinB-SH:
Re:
> Play around with "api" vs "implementation" dependencies. It's a concept introduced in gradle 4, I think. Basically API means that you export that API along with your library. But things like logging that isn't exposed publicly can be "implementation". See https://docs.gradle.org/current/userguide/java_library_plugin.html

Thank you for sending that link. I actually *need* to invest the time and learn `gradle` properly.

Here is on their website:
> The compile configuration still exists but should not be used as it will not offer the guarantees that the api and implementation configurations provide.

Why not depricate it then?!

**Comment by adomokos** [comment]
This is how the `UnfortunateException` is thrown and handled by the `Client`:

unfortunate-exception

**Comment by JustinB-SH** [comment]
@ben @adomokos RE: .gitignore the auto-generated files.  You can try to get thrift to generate the java files into the "build" directory.  With the protoc compiler, it generates the java files to `/model/build/generated/source/proto/main`.  That way they look like a build artifact, and get wiped out when you do a `gradle clean`, which is what you would do if you changed the proto definition and wanted to regenerate.  
Usually the gradle source generator plugin is smart enough to update the gradle compilation stage to let it know where these newly generated source files are, so it can compile those automatically.

**Comment by adomokos** [comment]
I'll do that in a new PR.

**Pull request merged**